scripts/git_changes_utils.py

```diff
@@ -0,0 +1,304 @@
+    if diff_filter:
+        git_cmd.append('--diff-filter={}'.format(diff_filter))
+    if left and right:
+        git_cmd.extend([left, right])
+        # Append -- to avoid conflicts between branch and directory name.
+        # More here: https://stackoverflow.com/questions/26349191
+        git_cmd.append('--')
+    out, err = common.start_subprocess_for_result(git_cmd)
+    if not err:
+        file_list = []
+        for line in out.splitlines():
+            # The lines in the output generally look similar to these:
+            #
+            #   A\tfilename
+            #   M\tfilename
+            #   R100\toldfilename\tnewfilename
+            #
+            # We extract the first char (indicating the status), and the string
+            # after the last tab character.
+            file_list.append(
+                FileDiff(bytes([line[0]]), line[line.rfind(b'\t') + 1:]))
+        return file_list
+    else:
+        raise ValueError(err)
+
+
+def get_merge_base(branch: str, other_branch: str) -> str:
+    """Returns the most-recent commit shared by both branches. Order doesn't
+    matter.
+
+    The commit returned is the same one used on GitHub's UI for comparing pull
+    requests.
+
+    Args:
+        branch: str. A branch name or commit hash.
+        other_branch: str. A branch name or commit hash.
+
+    Returns:
+        str. The common commit hash shared by both branches.
+
+    Raises:
+        ValueError. An error occurred in the git command.
+    """
+    merge_base, err = common.start_subprocess_for_result(
+        ['git', 'merge-base', branch, other_branch])
+    if err:
+        raise ValueError(err)
+
+    return merge_base.decode('utf-8').strip()
+
+
+def compare_to_remote(
+    remote: str, local_branch: str, remote_branch: Optional[str] = None
+) -> List[FileDiff]:
+    """Compare local with remote branch with git diff.
+
+    Parameter:
+        remote: str. Name of the git remote being pushed to.
+        local_branch: str. Name of the git branch being pushed to.
+        remote_branch: str|None. The name of the branch on the remote
+            to test against. If None, the remote branch is considered
+            to be the same as the local branch.
+
+    Returns:
+        list(FileDiff). List of FileDiffs that are modified, changed,
+        renamed or added but not deleted.
+
+    Raises:
+        ValueError. Raise ValueError if git command fails.
+    """
+    remote_branch = remote_branch if remote_branch else local_branch
+    git_remote = '%s/%s' % (remote, remote_branch)
+    # Ensure that references to the remote branches exist on the local machine.
+    common.start_subprocess_for_result(['git', 'pull', remote])
+    # Only compare differences to the merge base of the local and remote
+    # branches (what GitHub shows in the files tab of pull requests).
+    return git_diff_name_status(
+        get_merge_base(git_remote, local_branch), local_branch)
```

Comment:
Since we're getting this list of changes from a remote server that we shouldn't fully trust, we should validate it before relying on it. The only vulnerability I can see being applicable here is that the remote server might be able to get us to run checks on files outside of the oppia/ directory, which could lead to code outside oppia/ getting executed. To prevent this, can you check that every file returned by this function is inside the oppia/ directory?

Reply:
Sorry, can you elaborate on what exactly should be done here? When referring to checking if every file is inside the oppia/ directory do you just mean checking that it exists in the root oppia directory?

Reply:
This function is going to return a list of `FileDiff` objects. Before returning this list, make sure that each object in the list refers to a file inside the user's locally-cloned oppia/ directory. The point isn't to ensure that the file exists (in fact it might not since the file could have been deleted locally), but rather to ensure that the path points to inside the subtree rooted at oppia/.

For example, consider this file structure:

* `/`
  * `home/user`
    * `opensource/`
      * `oppia/`
        * `core/`
        * ...
      * `oppia_tools/`
      * ...
    * `dangerous_code/`
      * `danger.py`

Suppose that `danger.py` contains malicious code that shouldn't be executed (for example, maybe the developer was inspecting some malicious code for another project):

```python
import subprocess

subprocess.run(['someMaliciousCommand'])
```

We don't want to return a `FileDiff` object with the path `/home/user/dangerous_code/danger.py` or `../../dangerous_code/danger.py`, because that might cause a linter to import and therefore execute the dangerous code in `danger.py`. More broadly, this is an example of a common security principle that you should always validate potentially untrusted input to make sure it's safe

We can be pretty sure that code inside the `oppia/` directory is safe, so checking that every object in `FileDiff` points to somewhere inside `oppia/` should prevent this attack. Here's how you can check this:

```python
def check_file_inside_directory(file_path, directory_path):
    # os.path.abspath ensures all paths are absolute and eliminates things like `..` 
    # that make comparisons difficult.
    file_path = os.path.abspath(file_path)
    directory_path = os.path.abspath(directory_path)
    # If the file is inside the directory, then its path must include the full directory path.
    # Note that we don't just use `startswith()` or similar here because we don't want
    # to consider `/path/to/files` to be inside `/path/to/file`.
    return os.path.commonpath(file_path, directory_path) == directory_path
```

Reply:
Done, I didn't know what you meant by using the `oppia/` directory, I believe that meant the common.CURR_DIR?

Reply:
Yes, since we always run Oppia commands with the repository root (`oppia/`) as the current directory, this works